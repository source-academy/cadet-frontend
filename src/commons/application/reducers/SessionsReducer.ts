import { Reducer } from 'redux';

import { SourceActionType } from '../../utils/ActionsHelper';
import { defaultSession } from '../ApplicationTypes';
import {
  SET_GAME_STATE,
  UPDATE_MATERIAL_DIRECTORY_TREE,
  UPDATE_MATERIAL_INDEX
} from '../types/ActionTypes';
import { LOG_OUT } from '../types/CommonsTypes';
import {
  SessionState,
  SET_TOKENS,
  SET_USER,
  UPDATE_ASSESSMENT,
  UPDATE_ASSESSMENT_OVERVIEWS,
  UPDATE_GRADING,
  UPDATE_GRADING_OVERVIEWS,
  UPDATE_HISTORY_HELPERS,
  UPDATE_NOTIFICATIONS
} from '../types/SessionTypes';

export const SessionsReducer: Reducer<SessionState> = (
  state = defaultSession,
  action: SourceActionType
) => {
  switch (action.type) {
    case LOG_OUT:
      return defaultSession;
    case SET_TOKENS:
      return {
        ...state,
        accessToken: action.payload.accessToken,
        refreshToken: action.payload.refreshToken
      };
    case SET_USER:
      return {
        ...state,
        ...action.payload
      };
    case UPDATE_HISTORY_HELPERS:
      const helper = state.historyHelper;
      const isAcademy = isAcademyRe.exec(action.payload) != null;
      const newAcademyLocations = isAcademy
        ? [helper.lastAcademyLocations[1], action.payload]
        : [...helper.lastAcademyLocations];
      const newGeneralLocations = [helper.lastGeneralLocations[1], action.payload];
      return {
        ...state,
        historyHelper: {
          lastAcademyLocations: newAcademyLocations,
          lastGeneralLocations: newGeneralLocations
        }
      };
    case UPDATE_ASSESSMENT:
      const newAssessments = new Map(state.assessments);
      newAssessments.set(action.payload.id, action.payload);
      return {
        ...state,
        assessments: newAssessments
      };
    case UPDATE_ASSESSMENT_OVERVIEWS:
      return {
        ...state,
        assessmentOverviews: action.payload
      };
    case UPDATE_GRADING:
      const newGradings = new Map(state.gradings);
      newGradings.set(action.payload.submissionId, action.payload.grading);
      return {
        ...state,
        gradings: newGradings
      };
    case UPDATE_GRADING_OVERVIEWS:
      return {
        ...state,
        gradingOverviews: action.payload
      };
    case UPDATE_MATERIAL_DIRECTORY_TREE:
      return {
        ...state,
        materialDirectoryTree: action.payload.directoryTree
      };
    case UPDATE_MATERIAL_INDEX:
      return {
        ...state,
        materialIndex: action.payload.index
      };
    case UPDATE_NOTIFICATIONS:
      return {
        ...state,
        notifications: action.payload
      };
    case SET_GAME_STATE:
      return {
        ...state,
        gameState: action.payload
      };
    default:
      return state;
  }
};

export const isAcademyRe = new RegExp('^/academy.*', 'i');
