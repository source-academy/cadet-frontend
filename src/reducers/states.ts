import { Context } from 'js-slang'
import { SourceError } from 'js-slang/dist/types'

import { WorkspaceLocation, WorkspaceLocations } from '../actions/workspaces'
import { Grading, GradingOverview } from '../components/academy/grading/gradingShape'
import { Announcement } from '../components/Announcements'
import {
  ExternalLibraryName,
  ExternalLibraryNames,
  IAssessment,
  IAssessmentOverview
} from '../components/assessment/assessmentShape'
import { HistoryHelper } from '../utils/history'
import { createContext } from '../utils/slangHelper'

export interface IState {
  readonly academy: IAcademyState
  readonly application: IApplicationState
  readonly playground: IPlaygroundState
  readonly session: ISessionState
  readonly workspaces: IWorkspaceManagerState
}

export interface IAcademyState {
  readonly gameCanvas?: HTMLCanvasElement
}

export interface IApplicationState {
  readonly title: string
  readonly environment: ApplicationEnvironment
}

export interface IPlaygroundState {
  readonly queryString?: string
}

interface IAssessmentWorkspace extends IWorkspaceState {
  readonly currentAssessment?: number
  readonly currentQuestion?: number
  readonly hasUnsavedChanges: boolean
}

interface IGradingWorkspace extends IWorkspaceState {
  readonly currentSubmission?: number
  readonly currentQuestion?: number
  readonly hasUnsavedChanges: boolean
}

export interface IPlaygroundWorkspace extends IWorkspaceState {
  readonly playgroundExternal: ExternalLibraryName
}

export interface IWorkspaceManagerState {
  readonly assessment: IAssessmentWorkspace
  readonly grading: IGradingWorkspace
  readonly playground: IPlaygroundWorkspace
}

export interface IWorkspaceState {
  readonly context: Context
  readonly editorValue: string | null
  readonly editorWidth: string
  readonly isEditorAutorun: boolean
  readonly isRunning: boolean
  readonly output: InterpreterOutput[]
  readonly replHistory: ReplHistory
  readonly replValue: string
  readonly sideContentActiveTab: number
  readonly sideContentHeight?: number
  readonly globals: Array<[string, any]>
}

export interface ISessionState {
  readonly accessToken?: string
  readonly assessmentOverviews?: IAssessmentOverview[]
  readonly assessments: Map<number, IAssessment>
  readonly announcements?: Announcement[]
  readonly grade: number
  readonly gradingOverviews?: GradingOverview[]
  readonly gradings: Map<number, Grading>
  readonly historyHelper: HistoryHelper
  readonly maxGrade: number
  readonly maxXp: number
  readonly refreshToken?: string
  readonly role?: Role
  readonly story?: Story
  readonly name?: string
  readonly xp: number
}

type ReplHistory = {
  browseIndex: null | number // [0, 49] if browsing, else null
  records: string[]
}

export const maxBrowseIndex = 50

export type Story = {
  story: string
  playStory: boolean
}

/**
 * An output while the program is still being run in the interpreter. As a
 * result, there are no return values or SourceErrors yet. However, there could
 * have been calls to display (console.log) that need to be printed out.
 */
export type RunningOutput = {
  type: 'running'
  consoleLogs: string[]
}

/**
 * An output which reflects the program which the user had entered. Not a true
 * Output from the interpreter, but simply there to let he user know what had
 * been entered.
 */
export type CodeOutput = {
  type: 'code'
  value: string
}

/**
 * An output which represents a program being run successfully, i.e. with a
 * return value at the end. A program can have either a return value, or errors,
 * but not both.
 */
export type ResultOutput = {
  type: 'result'
  value: any
  consoleLogs: string[]
  runtime?: number
  isProgram?: boolean
}

/**
 * An output which represents a program being run unsuccessfully, i.e. with
 * errors at the end. A program can have either a return value, or errors, but
 * not both.
 */
export type ErrorOutput = {
  type: 'errors'
  errors: SourceError[]
  consoleLogs: string[]
}

export type InterpreterOutput = RunningOutput | CodeOutput | ResultOutput | ErrorOutput

export enum ApplicationEnvironment {
  Development = 'development',
  Production = 'production',
  Test = 'test'
}

export enum Role {
  Student = 'student',
  Staff = 'staff',
  Admin = 'admin'
}

/**
 * Defines what chapters are available for usage.
 * For external libraries, see externalLibraries.ts
 */
export const sourceChapters = [1, 2, 3, 4]
const latestSourceChapter = sourceChapters.slice(-1)[0]

const currentEnvironment = (): ApplicationEnvironment => {
  switch (process.env.NODE_ENV) {
    case 'development':
      return ApplicationEnvironment.Development
    case 'production':
      return ApplicationEnvironment.Production
    default:
      return ApplicationEnvironment.Test
  }
}

export const defaultAcademy: IAcademyState = {
  gameCanvas: undefined
}

export const defaultApplication: IApplicationState = {
  title: 'Cadet',
  environment: currentEnvironment()
}

export const defaultPlayground: IPlaygroundState = {}

export const defaultEditorValue = '// Type your program in here!'

/**
 * Create a default IWorkspaceState for 'resetting' a workspace.
 * Takes in parameters to set the js-slang library and chapter.
 *
 * @param location the location of the workspace, used for context
 */
export const createDefaultWorkspace = (location: WorkspaceLocation): IWorkspaceState => ({
  context: createContext<WorkspaceLocation>(latestSourceChapter, [], location),
  editorValue: location === WorkspaceLocations.playground ? defaultEditorValue : null,
  editorWidth: '50%',
  output: [],
  replHistory: {
    browseIndex: null,
    records: []
  },
  replValue: '',
  sideContentActiveTab: 0,
  globals: [],
  isEditorAutorun: false,
  isRunning: false
})

export const defaultComments = 'Comments **here**. Use `markdown` if you ~~are cool~~ want!'

export const defaultWorkspaceManager: IWorkspaceManagerState = {
  assessment: {
    ...createDefaultWorkspace(WorkspaceLocations.assessment),
    currentAssessment: undefined,
    currentQuestion: undefined,
    hasUnsavedChanges: false
  },
  grading: {
    ...createDefaultWorkspace(WorkspaceLocations.grading),
    currentSubmission: undefined,
    currentQuestion: undefined,
    hasUnsavedChanges: false
  },
  playground: {
    ...createDefaultWorkspace(WorkspaceLocations.playground),
    playgroundExternal: ExternalLibraryNames.NONE
  }
}

export const defaultSession: ISessionState = {
  accessToken: undefined,
  announcements: [
    {
      author: 'Aministrator',
      title: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
      content:
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam non tellus non ligula consectetur feugiat a at mauris. Ut sagittis, urna id suscipit dictum, ipsum diam sollicitudin enim, sed ultricies diam turpis vel libero. Maecenas finibus nisl ac lobortis ultricies. Integer molestie urna vitae nisi pharetra porttitor. Vestibulum massa diam, tristique quis ante a, posuere placerat magna. Pellentesque at lobortis purus, vitae imperdiet diam. Nulla eu rutrum neque. Aliquam efficitur consectetur ullamcorper. Pellentesque ultricies, diam ut vestibulum pellentesque, metus arcu laoreet ex, at mattis quam est non neque. Nam velit ipsum, posuere non porttitor commodo, lobortis in urna. Nulla facilisi. Donec mollis id nibh a luctus. Mauris vitae orci id velit pulvinar pellentesque non ut sapien. Curabitur eu consequat lorem. Proin pretium blandit scelerisque. Morbi ultricies, tellus non posuere pretium, felis sem convallis magna, ut sagittis elit felis et sem. Aliquam auctor suscipit condimentum. Nam posuere nulla dolor, in maximus risus feugiat vel. Phasellus vestibulum odio nec leo vehicula condimentum. Ut et semper libero. Cras fermentum mauris quis ex sodales, sit amet hendrerit augue lobortis. Maecenas eu dapibus enim, nec auctor est. Quisque quis erat erat. Curabitur sed rutrum felis, non venenatis diam. Fusce maximus rhoncus neque, in maximus velit semper eu. Fusce tempus lorem ut sodales pulvinar.',
      pinned: true
    }
  ],
  assessments: new Map<number, IAssessment>(),
  assessmentOverviews: undefined,
  grade: 0,
  gradingOverviews: undefined,
  gradings: new Map<number, Grading>(),
  historyHelper: {
    lastAcademyLocations: [null, null],
    lastGeneralLocations: [null, null]
  },
  maxGrade: 0,
  maxXp: 0,
  refreshToken: undefined,
  name: undefined,
  xp: 0
}

export const defaultState: IState = {
  academy: defaultAcademy,
  application: defaultApplication,
  playground: defaultPlayground,
  session: defaultSession,
  workspaces: defaultWorkspaceManager
}
