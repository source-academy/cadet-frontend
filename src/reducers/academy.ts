import { Reducer } from 'redux';

import { IAction, LOG_OUT, SAVE_CANVAS } from '../actions/actionTypes';
import { defaultAcademy, IAcademyState } from './states';

export const reducer: Reducer<IAcademyState> = (state = defaultAcademy, action: IAction) => {
  switch (action.type) {
    case LOG_OUT:
      return defaultAcademy;
    case SAVE_CANVAS:
      return {
        ...state,
        gameCanvas: action.payload
      };
    default:
      return state;
  }
};
