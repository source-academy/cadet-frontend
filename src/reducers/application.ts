import { Action, Reducer } from 'redux';
import { defaultApplication, IApplicationState } from './states';

export const reducer: Reducer<IApplicationState> = (state = defaultApplication, action: Action) => {
  return state;
};
