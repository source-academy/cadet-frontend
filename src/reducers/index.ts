import { reducer as academy } from './academy';
import { reducer as application } from './application';
import { reducer as playground } from './playground';
import { reducer as session } from './session';
import { reducer as workspaces } from './workspaces';

export default {
  academy,
  application,
  playground,
  session,
  workspaces
};
