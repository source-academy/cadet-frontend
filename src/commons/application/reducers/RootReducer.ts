import { connectRouter } from 'connected-react-router';
import { History } from 'history';
import { combineReducers } from 'redux';

import { AcademyReducer as academy } from '../../../features/academy/AcademyReducer';
import { DashboardReducer as dashboard } from '../../../features/dashboard/DashboardReducer';
import { PlaygroundReducer as playground } from '../../../features/playground/PlaygroundReducer';
import { WorkspaceReducer as workspaces } from '../../workspace/WorkspaceReducer';
import { ApplicationReducer as application } from '../ApplicationReducer';
import { SessionsReducer as session } from './SessionsReducer';

const createRootReducer = (history: History) =>
  combineReducers({
    router: connectRouter(history),
    academy,
    application,
    dashboard,
    playground,
    session,
    workspaces
  });

export default createRootReducer;
