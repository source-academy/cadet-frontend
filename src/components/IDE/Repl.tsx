import * as React from 'react'

import { Card } from '@blueprintjs/core'
import ReplControlContainer from '../../containers/IDE/ReplControlContainer'
import ReplInputContainer from '../../containers/IDE/ReplInputContainer'
import { InterpreterOutput } from '../../reducers/states'
import { parseError, toString } from '../../slang'

export interface IReplProps {
  output: InterpreterOutput[]
}

export interface IOutputProps {
  output: InterpreterOutput
}

const Repl: React.SFC<IReplProps> = props => {
  const cards = props.output.map((slice, index) => <Output output={slice} key={index} />)
  return (
    <div className="Repl">
      <div className="row">
        <ReplControlContainer />
      </div>
      {cards}
      <ReplInputContainer />
    </div>
  )
}

export const Output: React.SFC<IOutputProps> = props => {
  switch (props.output.type) {
    case 'result':
      return <Card>{toString(props.output.value)}</Card>
    case 'errors':
      return <Card>{parseError(props.output.errors)}</Card>
    case 'code':
      return <Card>{props.output.value}</Card>
    default:
      return <Card>''</Card>
  }
}

export default Repl
