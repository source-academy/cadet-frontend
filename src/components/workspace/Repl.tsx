import { Card } from '@blueprintjs/core';
import { parseError } from 'js-slang';
import { stringify } from 'js-slang/dist/interop';
import * as React from 'react';
import { HotKeys } from 'react-hotkeys';

import { InterpreterOutput } from '../../reducers/states';
import CanvasOutput from './CanvasOutput';
import ReplInput, { IReplInputProps } from './ReplInput';

export interface IReplProps {
  output: InterpreterOutput[];
  replValue: string;
  handleBrowseHistoryDown: () => void;
  handleBrowseHistoryUp: () => void;
  handleReplEval: () => void;
  handleReplValueChange: (newCode: string) => void;
}

export interface IOutputProps {
  output: InterpreterOutput;
}

class Repl extends React.PureComponent<IReplProps, {}> {
  public render() {
    const cards = this.props.output.map((slice, index) => <Output output={slice} key={index} />);
    const inputProps: IReplInputProps = this.props as IReplInputProps;
    return (
      <div className="Repl">
        <div className="repl-output-parent">
          {cards}
          <HotKeys className="repl-input-parent row pt-card pt-elevation-0" handlers={handlers}>
            <ReplInput {...inputProps} />
          </HotKeys>
        </div>
      </div>
    );
  }
}

export const Output: React.SFC<IOutputProps> = props => {
  switch (props.output.type) {
    case 'code':
      return (
        <Card>
          <pre className="codeOutput">{props.output.value}</pre>
        </Card>
      );
    case 'running':
      return (
        <Card>
          <pre className="logOutput">{props.output.consoleLogs.join('\n')}</pre>
        </Card>
      );
    case 'result':
      if (props.output.consoleLogs.length === 0) {
        return (
          <Card>
            <pre className="resultOutput">{renderResult(props.output.value)}</pre>
          </Card>
        );
      } else {
        return (
          <Card>
            <pre className="logOutput">{props.output.consoleLogs.join('\n')}</pre>
            <pre className="resultOutput">{renderResult(props.output.value)}</pre>
          </Card>
        );
      }
    case 'errors':
      if (props.output.consoleLogs.length === 0) {
        return (
          <Card>
            <pre className="errorOutput">{parseError(props.output.errors)}</pre>
          </Card>
        );
      } else {
        return (
          <Card>
            <pre className="logOutput">{props.output.consoleLogs.join('\n')}</pre>
            <br />
            <pre className="errorOutput">{parseError(props.output.errors)}</pre>
          </Card>
        );
      }
    default:
      return <Card>''</Card>;
  }
};

const renderResult = (value: any) => {
  /** A class which is the output of the show() function */
  const ShapeDrawn = (window as any).ShapeDrawn;
  if (typeof ShapeDrawn !== 'undefined' && value instanceof ShapeDrawn) {
    return <CanvasOutput />;
  } else {
    return stringify(value);
  }
};

/* Override handler, so does not trigger when focus is in editor */
const handlers = {
  goGreen: () => {}
};

export default Repl;
