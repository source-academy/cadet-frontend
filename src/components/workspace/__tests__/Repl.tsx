import { shallow } from 'enzyme'
import * as React from 'react'

import { ResultOutput } from '../../../reducers/states'
import Repl, { Output } from '../Repl'

test('Repl renders correctly', () => {
  const props = {
    output: [{ type: 'result', value: 'abc', consoleLogs: [] } as ResultOutput],
    replValue: '',
    handleReplValueChange: (newCode: string) => {},
    handleReplEval: () => {},
    handleReplOutputClear: () => {},
    handleChapterSelect: (e: React.ChangeEvent<HTMLSelectElement>) => {}
  }
  const app = <Repl {...props} />
  const tree = shallow(app)
  expect(tree.debug()).toMatchSnapshot()
})

test("Output renders correctly for InterpreterOutput.type === 'result'", () => {
  const props: ResultOutput = { type: 'result', value: 'def', consoleLogs: [] }
  const app = <Output {...{ output: props }} />
  const tree = shallow(app)
  expect(tree.debug()).toMatchSnapshot()
})
