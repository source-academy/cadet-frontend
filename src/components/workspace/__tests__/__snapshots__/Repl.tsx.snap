// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code output renders correctly 1`] = `
"<Blueprint3.Card elevation={0} interactive={false}>
  <Component className=\\"codeOutput\\">
    display(&#39;&#39;);
  </Component>
</Blueprint3.Card>"
`;

exports[`Empty output renders an empty card 1`] = `
"<Blueprint3.Card elevation={0} interactive={false}>
  &#39;&#39;
</Blueprint3.Card>"
`;

exports[`Error output (no consoleLogs) renders correctly 1`] = `
"<Blueprint3.Card elevation={0} interactive={false}>
  <Component className=\\"errorOutput\\">
    Line &lt;unknown&gt;: Expected , got .
  </Component>
</Blueprint3.Card>"
`;

exports[`Error output (with consoleLogs) renders correctly 1`] = `
"<Blueprint3.Card elevation={0} interactive={false}>
  <Component className=\\"logOutput\\">
    a
    bb
    cccccccccccccccccccccccccccccccc
    d
  </Component>
  <br />
  <Component className=\\"errorOutput\\">
    Line &lt;unknown&gt;: Expected , got .
  </Component>
</Blueprint3.Card>"
`;

exports[`Repl renders correctly 1`] = `
"<div className=\\"Repl\\" style={{...}}>
  <div className=\\"repl-output-parent\\">
    <Component output={{...}} usingSubst={false} />
    <Component output={{...}} usingSubst={false} />
    <Component output={{...}} usingSubst={false} />
    <Component output={{...}} usingSubst={false} />
    <HotKeys className=\\"repl-input-parent row bp3-card bp3-elevation-0\\" handlers={{...}}>
      <ReplInput handleBrowseHistoryDown={[Function: handleBrowseHistoryDown]} handleBrowseHistoryUp={[Function: handleBrowseHistoryUp]} handleReplValueChange={[Function: handleReplValueChange]} handleReplEval={[Function: handleReplEval]} handleReplOutputClear={[Function: handleReplOutputClear]} output={{...}} replValue=\\"\\" />
    </HotKeys>
  </div>
</div>"
`;

exports[`Result output (no consoleLogs) renders correctly 1`] = `
"<Blueprint3.Card elevation={0} interactive={false}>
  <Component className=\\"resultOutput\\">
    42
  </Component>
</Blueprint3.Card>"
`;

exports[`Result output (with consoleLogs) renders correctly 1`] = `
"<Blueprint3.Card elevation={0} interactive={false}>
  <Component className=\\"logOutput\\">
    a
    bb
    cccccccccccccccccccccccccccccccc
    d
  </Component>
  <Component className=\\"resultOutput\\">
    42
  </Component>
</Blueprint3.Card>"
`;

exports[`Running output renders correctly 1`] = `
"<Blueprint3.Card elevation={0} interactive={false}>
  <Component className=\\"logOutput\\">
    a
    bb
    cccccccccccccccccccccccccccccccc
    d
  </Component>
</Blueprint3.Card>"
`;
